### PR TITLE
Delete flattening 'branch' if leaf has empty select with no children

### LINKED ISSUE #523: Fix: Missing select in flatteningLookup
For the profile https://www.medizininformatik-initiative.de/fhir/ext/modul-patho/StructureDefinition/mii-pr-patho-report

the field `DiagnosticReport.extension:related-report.value[x]` has an empty select and no children.

```json
"DiagnosticReport.extension:related-report": {
        "parent": "DiagnosticReport.extension",
        "viewDefinition": {
          "forEachOrNull": "extension.where(url = 'http://hl7.org/fhir/StructureDefinition/workflow-relatedArtifact')",
          "select": []
        },
        "children": [
          "DiagnosticReport.extension:related-report.value[x]"
        ]
      },
      "DiagnosticReport.extension:related-report.value[x]": {
        "parent": "DiagnosticReport.extension:related-report",
        "viewDefinition": {
          "select": []
        }
      }
```

this leads to the following broken viewDefinition in aether:

```json
{
  "resourceType": "https://sql-on-fhir.org/ig/StructureDefinition/ViewDefinition",
  "name": "MII PR Patho Report",
  "status": "draft",
  "resource": "DiagnosticReport",
  "select": [
    {
      "column": [
        {
          "name": "id",
          "path": "id"
        },
        {
          "name": "patient",
          "path": "subject.reference"
        }
      ]
    },
    {
      "forEachOrNull": "extension.where(url = 'http://hl7.org/fhir/StructureDefinition/workflow-relatedArtifact')"
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by Frontman50**
flattening:
- LookupElements with empty select attribute and no children (or non-resolvable child ids) will be excluded from flatteningLookup.json.
- Deletion propagates upward until an element with columns or valid children is found
- Add integration test to detect empty selects in generated flatteningLookup
- Add unit test

### CODE REVIEW

File: flattening/core/flattening.py [hunk 1]
```diff
@@ -290,6 +292,44 @@ def flattening_post_process(
     return dict(sorted(res.items(), key=lambda item: item[0]))
 
 
+def filter_for_invalid_lookup(
+    element_id: str, lookup: Dict[str, FlatteningLookupElement]
+) -> Dict[str, FlatteningLookupElement]:
+    """
+    Returns {} when an element is deemed to be wrong, else returns the lookup.
```
Comment (paulolaup) [inline flattening/core/flattening.py:299]: What exactly is returned should be described in the `:return:` section. Don't use `{}` in the functions main description like this.
What does it mean for an element to be `wrong`? I would use a different descriptor here. Maybe something like `pruning leafless branches`. Using `wrong` kind of implies that there is faulty logic somewhere that this function addresses.

File: flattening/core/flattening.py [hunk 2]
```diff
@@ -290,6 +292,44 @@ def flattening_post_process(
...
+    :param element_id: id of the element, root of the lookup
+    :param lookup: flatteningLookup with element_id at root
+    :return: {} is of correct structure
+    """
+    lookup_element: FlatteningLookupElement = lookup.get(element_id)
+    if not lookup_element:
+        return {}
+
+    # if at least one referenced children exists
+    if lookup_element and lookup_element.children:
```
Comment (paulolaup) [inline flattening/core/flattening.py:315]: Shouldn't all referenced children exist? Ideally the logic would only include such children in the first place.

File: tests/integration/flattening/test_flattening.py [hunk 3]
```diff
@@ -231,3 +231,50 @@ def test_lookup_file_validity(flattening_lookup: Path, package_manager):
...
+    element in the lookup - otherwise the generated ``ViewDefinition`` ends up with a ``select`` entry that is
+    never filled (see issue with ``DiagnosticReport.extension:related-report.value[x]`` in
+    https://www.medizininformatik-initiative.de/fhir/ext/modul-patho/StructureDefinition/mii-pr-patho-report).
+    Elements using ``column`` instead of ``select`` (primitive leaves) are not affected by this rule, and neither
+    are elements whose ``select`` already carries content (e.g. primitively-typed polymorphic children).
+    """
+    errors = []
+    for elem_id, elem in lookup.elements.items():
+        vd = elem.view_definition
+        if vd is None or vd.select is None or len(vd.select) > 0:
```
Comment (paulolaup) [inline tests/integration/flattening/test_flattening.py:250]: What difference does it make whether `vd.select` is `None` or is an empty list?

**Review by paulolaup — CHANGES_REQUESTED** [review]

File: flattening/core/flattening.py [hunk 4]
```diff
@@ -1422,9 +1470,14 @@ def _flatten_element(
             else:
                 ext_flattening_lookup_elements = {}
 
-            flat_lookup_els.update({**res, **ext_flattening_lookup_elements})
+            if res:
+                flat_lookup_els.update(res)
+            if ext_flattening_lookup_elements:
+                flat_lookup_els.update(ext_flattening_lookup_elements)
+
+            # flat_lookup_els.update({**res, **ext_flattening_lookup_elements})
```
Comment (paulolaup) [inline flattening/core/flattening.py]: Remove commented out code

File: flattening/core/flattening.py [hunk 5]
```diff
@@ -290,6 +295,45 @@ def flattening_post_process(
     return dict(sorted(res.items(), key=lambda item: item[0]))
 
 
+def pruning_leafless_branches(
```
Comment (paulolaup) [inline flattening/core/flattening.py]: Rename to `prune_leafless_branches`

File: flattening/core/flattening.py (hunk 2 above)
Comment (paulolaup) [reply flattening/core/flattening.py]: I don't think that quite addresses my question. Is there no way to ensure that no such branches get generated in the first place? I think what I don't quite understand yet is when and why do get generated.
I think we should figure out why this happens and then determine what the best action is.

Also: You don't need to check if `lookup_element` exists from here on out since you already did that before

File: flattening/core/flattening.py [hunk 6]
```diff
@@ -541,7 +590,7 @@ def _flatten_coding(
                     select=[],
                 )
 
-                flat_element.children = [
+                children: List[str] = [
```
Comment (paulolaup) [inline flattening/core/flattening.py]: With the introduction if the new structure definition data model you can probably simply the logic here

File: tests/integration/flattening/test_flattening.py (hunk 3 above)
Comment (paulolaup) [reply tests/integration/flattening/test_flattening.py]: What about lookup elements with `view_definition.column`? Shouldn't they also be valid leafs?

**Review by paulolaup — CHANGES_REQUESTED** [review]

File: flattening/core/flattening.py (hunk 2 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: Leafless branches get generated whenever the flattening lookup generation encounters unsupported or excluded types. Predicting branches with such leafs and thus avoid generation of leafless branches, would require rethinking approach to work the other way around, going from leaves -> root. The cost outweighs the benefit of the small performance increase, imo.

File: tests/integration/flattening/test_flattening.py (hunk 3 above)
Comment (Frontman50) [reply tests/integration/flattening/test_flattening.py]: "`vd.select is None`" implies the case when `view_definition.column` is defined, since these are by definition mutually exclusive. 

https://github.com/medizininformatik-initiative/fhir-ontology-generator/blob/652aa3241b713707b90c64715505a7514655226c/flattening/model/FlatteningLookupModels.py#L30

File: common/util/fhirpath/functions.py [hunk 7]
```diff
@@ -23,11 +23,439 @@
...
+    slicing_def_elem_def = slice_elem_def.parent
+    if (
+        slicing_def_elem_def.type
+        and len(slicing_def_elem_def.type) == 1
+        and slicing_def_elem_def.type[0].code == "Reference"
+    ):
+        # If the slicing-defining element definition supports only the data type 'Reference' then the type
+        # discriminator is applied to the resource type of the reference target
+        target_types = []
+        for t_pr in slicing_def_elem_def.type[0].targetProfile:
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: Shouldn't it be `slice_elem_def`? We need the profiles allowed for this slice to generate the filter. Probably an oversight by me

File: common/util/fhirpath/functions.py [hunk 8]
```diff
@@ -23,11 +23,439 @@
...
+    (e.g. `Coding.system`), without repeating that constraint locally as a child element definition of the slice.
+    Resolving that referenced profile is required to find the discriminator value in such cases
+
+    :param elem_def: Element definition whose type profile should be resolved
+    :param manager: FHIR package manager used to resolve the profile, or `None` if unavailable
+    :return: Root element definition (e.g. `Coding`) of the resolved profile, or `None` if it could not be resolved
+        (no manager available, no single unambiguous type profile, or the profile could not be found)
+    """
+    if not manager or not elem_def.type:
+        return None
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: Raise Exception or Error

File: common/util/fhirpath/functions.py [hunk 9]
```diff
@@ -23,11 +23,439 @@
...
+    profile_urls = {p for t in elem_def.type for p in (t.profile or [])}
+    if len(profile_urls) != 1:
+        return None
+    resolved = manager.find_struct_def(next(iter(profile_urls)))
+    if not resolved:
+        return None
+    return next((e for e in resolved.snapshot.element if e.id == resolved.type), None)
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: Use ``NavStructureDefinition.root``

File: common/util/fhirpath/functions.py [hunk 10]
```diff
@@ -23,11 +23,439 @@
...
+    Resolving that referenced profile is required to find the discriminator value in such cases
+
+    :param elem_def: Element definition whose type profile should be resolved
+    :param manager: FHIR package manager used to resolve the profile, or `None` if unavailable
+    :return: Root element definition (e.g. `Coding`) of the resolved profile, or `None` if it could not be resolved
+        (no manager available, no single unambiguous type profile, or the profile could not be found)
+    """
+    if not manager or not elem_def.type:
+        return None
+    profile_urls = {p for t in elem_def.type for p in (t.profile or [])}
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: Do you want to do this over all types? E.g. what if there are two data types that the element definition supports and each one references a dedicated specialization of itself (profile)?

File: common/util/fhirpath/functions.py [hunk 11]
```diff
@@ -23,11 +23,439 @@
...
+    if (
+        len(slice_elem_def.type) == 1
+        and slice_elem_def.type[0].code == "Extension"
+        and discr_path == "url"
+    ):
+        return fhirpath_filter_for_profile_discriminated_slice(
+            slice_elem_def, "$this", _criteria_only
+        )
+
+    # The discriminator path only identifies which fixed/pattern value distinguishes this slice from its
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: I don't think thats right. If there is a discriminator path then all fixed/pattern/binding values at that path (or further down) could potentially be required to distinguish the slice from the others

File: common/util/fhirpath/functions.py [hunk 12]
```diff
@@ -23,11 +23,439 @@
...
+        parent_elem_def.slicing.discriminator
+        if parent_elem_def and parent_elem_def.slicing
+        else []
+    )
+    if not discriminators:
+        raise ValueError(
+            f"Parent element definition '{parent_elem_def.id}' defines no slicing discriminators"
+        )
+    exprs = []
+    # A "value"/"pattern" discriminator always resolves to the same combined expression regardless of which
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: No that shouldn't be the case

File: common/util/fhirpath/functions.py [hunk 13]
```diff
@@ -327,30 +755,112 @@ def _append_filter_from_profile_discriminated_elem(
...
-                    f"{'' if discr_path else (elem_def.rel_path + '.')}memberOf('{binding.valueSet}')"
-                )
+            if is_binding_target and binding and binding.valueSet:
+                target = "$this" if discr_path else elem_def.rel_path
+                binding_expr = _filter_from_binding(elem_def, binding, target, client)
+                if binding_expr:
+                    if binding.strength == "required":
+                        exprs.append(binding_expr)
+                    else:
+                        # A non-required binding is not trusted outright -- it is only a last-resort fallback, used
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: non-required -> optional

File: common/util/fhirpath/functions.py [hunk 14]
```diff
@@ -609,7 +1153,48 @@ def filter_for_slice(
             case "exists":
                 exprs.append(f"{discr_path}.exists()")
             case "type":
-                if single_discriminator and len(slice_elem_def.type) == 1:
+                reference_types = [
+                    t
+                    for t in slice_elem_def.type
+                    if t.code in ("Reference", "canonical")
+                ]
+                if reference_types and "resolve()" in discr_path:
```
Comment (paulolaup) [inline common/util/fhirpath/functions.py]: Isn't this case already covered by a previous function? Why not use it instead. Looks like code duplication to me

File: flattening/core/flattening.py [hunk 15]
```diff
@@ -264,18 +268,61 @@ def recontextualize_extension_lookup(
...
+        3. Removes empty children arrays
+        4. Removes references to non-existing children
+        5. Sorts entries by index to enable easy comparison between versions
     :param lookup_elements: Generated Lookup
     :return: Post processed lookup
     """
+    # 1 Prunes leafless branches and 2 Removes elements and branches with "resolve()"
+    for key in reversed(list(lookup_elements.keys())):
+        el: FlatteningLookupElement = lookup_elements.get(key)
+        if not el:
```
Comment (paulolaup) [inline flattening/core/flattening.py]: Should never happen. Why would there be keys with value `None`?

File: flattening/core/flattening.py [hunk 16]
```diff
@@ -264,18 +268,61 @@ def recontextualize_extension_lookup(
...
+
+            # delete upward until parent is root or without any valid children
+            parent_key: str = key
+            while (
+                parent_key := flattening_get_parent(parent_key, lookup_elements)
+            ) and len(
+                [
+                    sibling_id
+                    for sibling_id in lookup_elements.get(parent_key).children
+                    if lookup_elements.get(sibling_id)
```
Comment (paulolaup) [inline flattening/core/flattening.py]: Same as above

File: flattening/core/flattening.py [hunk 17]
```diff
@@ -287,9 +334,84 @@ def flattening_post_process(
...
+    :param element_id: id of the element, root of the lookup
+    :param lookup: flatteningLookup with element_id at root
+    :return: {} is of correct structure
+    """
+    lookup_element: FlatteningLookupElement | None = lookup.get(element_id)
+    if not lookup_element:
+        return False
+
+    # if at least one referenced children exists
+    if lookup_element and lookup_element.children:
```
Comment (paulolaup) [inline flattening/core/flattening.py]: You ensured that `lookup_element` exists

File: flattening/core/flattening.py [hunk 18]
```diff
@@ -349,6 +471,66 @@ def _get_required_children_for_element(
             for child_spec in self.config.required_children_per_element.get(type, [])
         ]
 
+    @staticmethod
+    def _log_discriminator_issue(
+        profile: StructureDefinitionSnapshot, message: str
+    ) -> None:
+        """Logs a dropped-slice reason: warning for a draft profile, error otherwise."""
+        log = _logger.warning if profile.status == "draft" else _logger.error
```
Comment (paulolaup) [inline flattening/core/flattening.py]: Profiles with status `draft` shouldn't even be processed in the first place

File: tests/integration/conftest.py [hunk 19]
```diff
@@ -21,9 +22,12 @@ def pytest_addoption(parser):
 
 
 @pytest.fixture(scope="session", autouse=True)
-def copy_and_unpack_project_output(pytestconfig) -> Path:
-    project = Project(name=pytestconfig.getoption("--project"))
+def project(pytestconfig) -> Project:
```
Comment (paulolaup) [inline tests/integration/conftest.py]: That fixture should already be available if I'm not mistaken

File: tests/unit/flattening/builders.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    Builds a minimal navigable profile snapshot: a synthetic root element for
+    ``resource_type`` plus whatever ``elements`` the test needs.
+
+    :param resource_type: FHIR resource type of the profile (e.g. "Observation")
+    :param elements: element definitions to include, in addition to the root element
+    :param url: canonical URL of the profile, defaults to a synthetic example.org URL
+    :return: navigable profile snapshot usable wherever a ``StructureDefinitionSnapshot`` is expected
+    """
+    root = ElementDefinition(id=resource_type, path=resource_type)
+    struct_def = StructureDefinition(
```
Comment (paulolaup) [inline tests/unit/flattening/builders.py]: You can directly construct an instance of `NavStructureDefinition`

File: tests/unit/flattening/test_flattening.py [hunk 21]
```diff
@@ -56,9769 +55,2401 @@
...
         ),
-        (
-            "https://www.medizininformatik-initiative.de/fhir/core/modul-diagnose/StructureDefinition/Diagnose",
-            "Condition.code.coding:icd10-gm.system",
+        pytest.param(
+            # Regression for the "Account.coverage.extension:Abrechnungsart.value[x]:valueCoding"
+            # case: a polymorphic child whose columns are wrapped inside `select` (as produced by
+            # `_generate_flattening_polymorphic_child`) instead of a top-level `column`, and which
+            # has no `children` of its own. This used to be wrongly treated as "nothing to
+            # contribute" and dropped, which cascaded upward and deleted the whole extension chain.
```
Comment (paulolaup) [inline tests/unit/flattening/test_flattening.py]: What is meant by "nothing to contribute"?

File: tests/unit/flattening/test_flattening.py [hunk 22]
```diff
@@ -56,9769 +55,2401 @@
...
-                    view_definition=ViewDefinitionSnippet(
-                        column=[
-                            ViewDefinitionColumn(
-                                name="Task_input_value_X_Valuetiming_repeat_bounds_X_Boundsrange_high_unit",
-                                path="unit",
-                                type="string",
-                            ),
-                        ],
-                    ),
-
```
Comment (paulolaup) [inline tests/unit/flattening/test_flattening.py]: I find the test name a bit confusing. The description explains what is tested very well though.

File: flattening/core/flattening.py (hunk 15 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: because the elements of the lookup are actively beeing deleted. This means that it could be that key points to a child that was already deleted. To prevent this the list of keys is reversed, so that children get processed before their parents, but it does not hurt to prevent a crash

File: flattening/core/flattening.py (hunk 16 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: again, ... it could be that the sibling is already deleted

File: flattening/core/flattening.py (hunk 17 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: I know, but pycharm is complaining

File: flattening/core/flattening.py (hunk 18 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: Changed to considering only profiles with status active for flattening

File: tests/unit/flattening/builders.py (hunk 20 above)
Comment (Frontman50) [reply tests/unit/flattening/builders.py]: Fix applied

File: tests/unit/flattening/test_flattening.py (hunk 21 above)
Comment (Frontman50) [reply tests/unit/flattening/test_flattening.py]: "Leafless branch" is meant here. Bad wording

File: tests/integration/conftest.py (hunk 19 above)
Comment (Frontman50) [reply tests/integration/conftest.py]: removed fixture

File: common/util/fhirpath/functions.py (hunk 12 above)
Comment (Frontman50) [reply common/util/fhirpath/functions.py]: reworked function

File: flattening/core/flattening.py (hunk 1 above)
Comment (Frontman50) [reply flattening/core/flattening.py]: Changed wording

File: flattening/core/flattening.py (hunk 15 above)
Comment (paulolaup) [reply flattening/core/flattening.py]: I'd be better if we remove this check since if it were to crash without it we would know that there are still some issue with the implementation

File: flattening/core/flattening.py (hunk 17 above)
Comment (paulolaup) [reply flattening/core/flattening.py]: It is annoying I know but that should not lead to us adding unnecessary checks

File: tests/integration/flattening/test_flattening.py (hunk 3 above)
Comment (paulolaup) [reply tests/integration/flattening/test_flattening.py]: Ok

File: tests/resources/project/package.json [hunk 23]
```diff
@@ -14,18 +14,20 @@
     "de.ihe-d.terminology": "3.0.1",
     "de.medizininformatikinitiative.kerndatensatz.bildgebung": "2025.0.2",
     "de.medizininformatikinitiative.kerndatensatz.biobank": "2025.0.4",
-    "de.medizininformatikinitiative.kerndatensatz.consent": "2025.0.4",
-    "de.medizininformatikinitiative.kerndatensatz.diagnose": "2025.0.0",
-    "de.medizininformatikinitiative.kerndatensatz.fall": "2025.0.0",
+    "de.medizininformatikinitiative.kerndatensatz.consent": "2026.0.0",
```
Comment (paulolaup) [inline tests/resources/project/package.json]: Why did you switch to the 2026 version?

**Review by paulolaup — CHANGES_REQUESTED** [review]

**Review by paulolaup — APPROVED** [review]
